From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: thongass000 <tienthong2004@gmail.com>
Date: Wed, 7 Oct 2026 00:53:45 +0700
Subject: [PATCH 5/7] EpdgService: Run the IWLAN data call on the eris tunnel

Samsung's kernel has virtual rmnet<n> interfaces that carry each IWLAN
data call, and its netd redirects them into the ePDG tunnel. This kernel
has neither, so report the data call on the tunnel itself (epdg<n>, the
eris TUN device) and configure its addresses there; EpdgCompat hands
its IPv6 addresses to the device's epdg_addr helper, since
setInterfaceConfig() only sets IPv4. Interface events for epdg<n> are
passed on as rmnet<n>, the name EpdgService tracks the data call by,
and the tunnel's link down events are dropped: it is down while eris
brings it up, and a lost tunnel is reported by eris itself.

Also implement interfaceClassDataActivityChanged(int, boolean, long,
int), which Android 12's INetworkManagementEventObserver added.

Assisted-by: LLM
---
 .../sec/epdg/EpdgNetworkMgmtObserver.smali    | 35 ++++++++++++++++
 ...iption$EpdgNetworkManagementObserver.smali | 41 ++++++++++++++-----
 smali/com/sec/epdg/TelephonyAdapter.smali     |  2 +-
 .../interfaceController/EpdgInterface.smali   | 21 ++++++++++
 ...faceController$EpdgInterfaceObserver.smali | 41 ++++++++++++++-----
 5 files changed, 119 insertions(+), 21 deletions(-)

diff --git a/smali/com/sec/epdg/EpdgNetworkMgmtObserver.smali b/smali/com/sec/epdg/EpdgNetworkMgmtObserver.smali
index 24c0afd..daf45e2 100644
--- a/smali/com/sec/epdg/EpdgNetworkMgmtObserver.smali
+++ b/smali/com/sec/epdg/EpdgNetworkMgmtObserver.smali
@@ -46,6 +46,35 @@
     return-void
 .end method
 
+.method protected static toMobileIface(Ljava/lang/String;)Ljava/lang/String;
+    .locals 2
+
+    const-string v0, "epdg"
+
+    invoke-virtual {p0, v0}, Ljava/lang/String;->startsWith(Ljava/lang/String;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_0
+
+    invoke-static {}, Lcom/sec/epdg/EpdgUtils;->getMobileInterfacePrefix()Ljava/lang/String;
+
+    move-result-object v0
+
+    const/4 v1, 0x4
+
+    invoke-virtual {p0, v1}, Ljava/lang/String;->substring(I)Ljava/lang/String;
+
+    move-result-object v1
+
+    invoke-virtual {v0, v1}, Ljava/lang/String;->concat(Ljava/lang/String;)Ljava/lang/String;
+
+    move-result-object p0
+
+    :cond_0
+    return-object p0
+.end method
+
 
 # virtual methods
 .method public EpdgAddressRemoved(Landroid/net/LinkAddress;Ljava/lang/String;)V
@@ -232,6 +261,12 @@
     return-void
 .end method
 
+.method public interfaceClassDataActivityChanged(IZJI)V
+    .locals 0
+
+    return-void
+.end method
+
 .method public interfaceClassDataActivityChanged(Ljava/lang/String;ZJ)V
     .locals 2
     .param p1, "label"    # Ljava/lang/String;
diff --git a/smali/com/sec/epdg/EpdgSubScription$EpdgNetworkManagementObserver.smali b/smali/com/sec/epdg/EpdgSubScription$EpdgNetworkManagementObserver.smali
index e85136e..cc106ba 100644
--- a/smali/com/sec/epdg/EpdgSubScription$EpdgNetworkManagementObserver.smali
+++ b/smali/com/sec/epdg/EpdgSubScription$EpdgNetworkManagementObserver.smali
@@ -40,6 +40,10 @@
     .param p1, "linkAddress"    # Landroid/net/LinkAddress;
     .param p2, "iface"    # Ljava/lang/String;
 
+    invoke-static {p2}, Lcom/sec/epdg/EpdgNetworkMgmtObserver;->toMobileIface(Ljava/lang/String;)Ljava/lang/String;
+
+    move-result-object p2
+
     .line 3912
     if-nez p1, :cond_0
 
@@ -367,6 +371,23 @@
     .param p1, "iface"    # Ljava/lang/String;
     .param p2, "up"    # Z
 
+    const-string v0, "epdg"
+
+    invoke-virtual {p1, v0}, Ljava/lang/String;->startsWith(Ljava/lang/String;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_0
+
+    if-nez p2, :cond_0
+
+    return-void
+
+    :cond_0
+    invoke-static {p1}, Lcom/sec/epdg/EpdgNetworkMgmtObserver;->toMobileIface(Ljava/lang/String;)Ljava/lang/String;
+
+    move-result-object p1
+
     .line 3863
     iget-object v0, p0, Lcom/sec/epdg/EpdgSubScription$EpdgNetworkManagementObserver;->this$0:Lcom/sec/epdg/EpdgSubScription;
 
@@ -388,13 +409,13 @@
 
     invoke-virtual {v1, v2}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
 
-    if-eqz p2, :cond_0
+    if-eqz p2, :cond_1
 
     const-string v2, "up"
 
     goto :goto_0
 
-    :cond_0
+    :cond_1
     const-string v2, "down"
 
     :goto_0
@@ -464,7 +485,7 @@
 
     move-result v5
 
-    if-gt v4, v5, :cond_1
+    if-gt v4, v5, :cond_2
 
     .line 3871
     iget-object v4, p0, Lcom/sec/epdg/EpdgSubScription$EpdgNetworkManagementObserver;->this$0:Lcom/sec/epdg/EpdgSubScription;
@@ -481,7 +502,7 @@
     return-void
 
     .line 3875
-    :cond_1
+    :cond_2
     const/4 v4, 0x0
 
     :try_start_0
@@ -600,16 +621,16 @@
 
     move-result v4
 
-    if-eqz v4, :cond_4
+    if-eqz v4, :cond_5
 
     sget v4, Lcom/sec/epdg/IWlanEnum;->IWLAN_MAX_SETTING:I
 
-    if-lt v3, v4, :cond_2
+    if-lt v3, v4, :cond_3
 
     goto :goto_2
 
     .line 3894
-    :cond_2
+    :cond_3
     iget-object v4, p0, Lcom/sec/epdg/EpdgSubScription$EpdgNetworkManagementObserver;->this$0:Lcom/sec/epdg/EpdgSubScription;
 
     invoke-static {v4}, Lcom/sec/epdg/EpdgSubScription;->access$7600(Lcom/sec/epdg/EpdgSubScription;)[Lcom/sec/epdg/IPSecDataConnSM;
@@ -630,7 +651,7 @@
 
     .line 3896
     .local v4, "msgToHandoverSM":Landroid/os/Message;
-    if-nez p2, :cond_3
+    if-nez p2, :cond_4
 
     .line 3897
     iget-object v5, p0, Lcom/sec/epdg/EpdgSubScription$EpdgNetworkManagementObserver;->this$0:Lcom/sec/epdg/EpdgSubScription;
@@ -651,7 +672,7 @@
     goto :goto_1
 
     .line 3901
-    :cond_3
+    :cond_4
     iget-object v5, p0, Lcom/sec/epdg/EpdgSubScription$EpdgNetworkManagementObserver;->this$0:Lcom/sec/epdg/EpdgSubScription;
 
     invoke-static {v5}, Lcom/sec/epdg/EpdgSubScription;->access$000(Lcom/sec/epdg/EpdgSubScription;)Ljava/lang/String;
@@ -680,7 +701,7 @@
 
     .line 3892
     .end local v4    # "msgToHandoverSM":Landroid/os/Message;
-    :cond_4
+    :cond_5
     :goto_2
     iget-object v4, p0, Lcom/sec/epdg/EpdgSubScription$EpdgNetworkManagementObserver;->this$0:Lcom/sec/epdg/EpdgSubScription;
 
diff --git a/smali/com/sec/epdg/TelephonyAdapter.smali b/smali/com/sec/epdg/TelephonyAdapter.smali
index b879136..0e05e54 100644
--- a/smali/com/sec/epdg/TelephonyAdapter.smali
+++ b/smali/com/sec/epdg/TelephonyAdapter.smali
@@ -1563,7 +1563,7 @@
 
     sub-int/2addr v4, v8
 
-    invoke-virtual {v3, v4}, Lcom/sec/epdg/EpdgUtils;->getMobileInterfaceName(I)Ljava/lang/String;
+    invoke-virtual {v3, v4}, Lcom/sec/epdg/EpdgUtils;->getIwlanInterfaceName(I)Ljava/lang/String;
 
     move-result-object v3
 
diff --git a/smali/com/sec/epdg/interfaceController/EpdgInterface.smali b/smali/com/sec/epdg/interfaceController/EpdgInterface.smali
index c8bbb2c..f448118 100644
--- a/smali/com/sec/epdg/interfaceController/EpdgInterface.smali
+++ b/smali/com/sec/epdg/interfaceController/EpdgInterface.smali
@@ -149,7 +149,18 @@
     .locals 3
     .param p1, "radioIface"    # Ljava/lang/String;
 
+    const-string v0, "epdg"
+
+    invoke-virtual {p1, v0}, Ljava/lang/String;->startsWith(Ljava/lang/String;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_0
+
+    return-object p1
+
     .line 67
+    :cond_0
     const-string v0, "epdg"
 
     .line 68
@@ -195,6 +206,10 @@
     .locals 3
     .param p1, "iface"    # Ljava/lang/String;
 
+    invoke-direct {p0, p1}, Lcom/sec/epdg/interfaceController/EpdgInterface;->getTunInterface(Ljava/lang/String;)Ljava/lang/String;
+
+    move-result-object p1
+
     .line 142
     iget-object v0, p0, Lcom/sec/epdg/interfaceController/EpdgInterface;->TAG:Ljava/lang/String;
 
@@ -297,6 +312,10 @@
 
     .line 106
     .local p2, "addresses":Ljava/util/List;, "Ljava/util/List<Landroid/net/LinkAddress;>;"
+    invoke-direct {p0, p1}, Lcom/sec/epdg/interfaceController/EpdgInterface;->getTunInterface(Ljava/lang/String;)Ljava/lang/String;
+
+    move-result-object p1
+
     iget-object v0, p0, Lcom/sec/epdg/interfaceController/EpdgInterface;->TAG:Ljava/lang/String;
 
     new-instance v1, Ljava/lang/StringBuilder;
@@ -387,6 +406,8 @@
 
     .line 131
     :goto_0
+    invoke-static {p1, p2}, Lcom/sec/epdg/compat/EpdgCompat;->addTunnelAddresses(Ljava/lang/String;Ljava/util/List;)V
+
     iget-object v3, p0, Lcom/sec/epdg/interfaceController/EpdgInterface;->TAG:Ljava/lang/String;
 
     new-instance v4, Ljava/lang/StringBuilder;
diff --git a/smali/com/sec/epdg/interfaceController/EpdgInterfaceController$EpdgInterfaceObserver.smali b/smali/com/sec/epdg/interfaceController/EpdgInterfaceController$EpdgInterfaceObserver.smali
index 2c07315..05756ad 100644
--- a/smali/com/sec/epdg/interfaceController/EpdgInterfaceController$EpdgInterfaceObserver.smali
+++ b/smali/com/sec/epdg/interfaceController/EpdgInterfaceController$EpdgInterfaceObserver.smali
@@ -47,6 +47,10 @@
     .param p1, "linkAddress"    # Landroid/net/LinkAddress;
     .param p2, "iface"    # Ljava/lang/String;
 
+    invoke-static {p2}, Lcom/sec/epdg/EpdgNetworkMgmtObserver;->toMobileIface(Ljava/lang/String;)Ljava/lang/String;
+
+    move-result-object p2
+
     .line 233
     iget-object v0, p0, Lcom/sec/epdg/interfaceController/EpdgInterfaceController$EpdgInterfaceObserver;->this$0:Lcom/sec/epdg/interfaceController/EpdgInterfaceController;
 
@@ -288,6 +292,23 @@
     .param p1, "iface"    # Ljava/lang/String;
     .param p2, "up"    # Z
 
+    const-string v0, "epdg"
+
+    invoke-virtual {p1, v0}, Ljava/lang/String;->startsWith(Ljava/lang/String;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_0
+
+    if-nez p2, :cond_0
+
+    return-void
+
+    :cond_0
+    invoke-static {p1}, Lcom/sec/epdg/EpdgNetworkMgmtObserver;->toMobileIface(Ljava/lang/String;)Ljava/lang/String;
+
+    move-result-object p1
+
     .line 187
     iget-object v0, p0, Lcom/sec/epdg/interfaceController/EpdgInterfaceController$EpdgInterfaceObserver;->INTTAG:Ljava/lang/String;
 
@@ -305,13 +326,13 @@
 
     invoke-virtual {v1, v2}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
 
-    if-eqz p2, :cond_0
+    if-eqz p2, :cond_1
 
     const-string v2, "up"
 
     goto :goto_0
 
-    :cond_0
+    :cond_1
     const-string v2, "down"
 
     :goto_0
@@ -358,7 +379,7 @@
 
     move-result v5
 
-    if-gt v4, v5, :cond_1
+    if-gt v4, v5, :cond_2
 
     .line 194
     iget-object v4, p0, Lcom/sec/epdg/interfaceController/EpdgInterfaceController$EpdgInterfaceObserver;->this$0:Lcom/sec/epdg/interfaceController/EpdgInterfaceController;
@@ -375,7 +396,7 @@
     return-void
 
     .line 198
-    :cond_1
+    :cond_2
     :try_start_0
     invoke-virtual {v0}, Ljava/lang/String;->length()I
 
@@ -463,16 +484,16 @@
 
     move-result v4
 
-    if-eqz v4, :cond_4
+    if-eqz v4, :cond_5
 
     sget v4, Lcom/sec/epdg/IWlanEnum;->IWLAN_MAX_SETTING:I
 
-    if-lt v3, v4, :cond_2
+    if-lt v3, v4, :cond_3
 
     goto :goto_2
 
     .line 216
-    :cond_2
+    :cond_3
     iget-object v4, p0, Lcom/sec/epdg/interfaceController/EpdgInterfaceController$EpdgInterfaceObserver;->this$0:Lcom/sec/epdg/interfaceController/EpdgInterfaceController;
 
     invoke-static {v4}, Lcom/sec/epdg/interfaceController/EpdgInterfaceController;->access$000(Lcom/sec/epdg/interfaceController/EpdgInterfaceController;)I
@@ -489,7 +510,7 @@
 
     .line 217
     .local v4, "interfaceMsg":Landroid/os/Message;
-    if-nez p2, :cond_3
+    if-nez p2, :cond_4
 
     .line 218
     iget-object v6, p0, Lcom/sec/epdg/interfaceController/EpdgInterfaceController$EpdgInterfaceObserver;->INTTAG:Ljava/lang/String;
@@ -504,7 +525,7 @@
     goto :goto_1
 
     .line 222
-    :cond_3
+    :cond_4
     iget-object v5, p0, Lcom/sec/epdg/interfaceController/EpdgInterfaceController$EpdgInterfaceObserver;->INTTAG:Ljava/lang/String;
 
     const-string v6, "Mobile Inteface is up"
@@ -529,7 +550,7 @@
 
     .line 214
     .end local v4    # "interfaceMsg":Landroid/os/Message;
-    :cond_4
+    :cond_5
     :goto_2
     iget-object v4, p0, Lcom/sec/epdg/interfaceController/EpdgInterfaceController$EpdgInterfaceObserver;->INTTAG:Ljava/lang/String;
 
